### PR TITLE
Devel/ignore no card

### PR DESCRIPTION
**Pull request by wolneykien**
Don't stuck if `wait_for_card=false` and ignore the token not found error when the auth isn't restricted to card only (either by option or by `PKCS11_LOGIN_TOKEN_NAME` environment variable).

It allows to pass to a next module in the PAM stack if the auth isn't restricted to card only. For instance:

```
auth		[success=done ignore=ignore default=die]	pam_pkcs11.so
auth		requisite	pam_succeed_if.so user ingroup wheel
auth		include		system-auth-local
```

The meaning of `card_only` flags changes to the following: restrict the authentication to token _only if_ the user has
already authenticated by a token. Thus, in order to restrict the authentication for _new logins_ to card only the `wait_for_card` flag should be used _along_ with `card_only` flag.

### CODE REVIEW

**Comment by wolneykien** [comment]
Please, take a look over this PR. It changes `card_only` logic for new logins: it returns `PAM_IGNORE` if no suitable token was found. Why? The system administrator should configure `/etc/pam.d` files in order to limit the available authentication methods to `pam_pkcs11` or to provide some additional methods. The `pam_pkcs11` module is not the right place to make decisions about the available authentication methods.

However, for existing logins (i.e. `PKCS11_LOGIN_TOKEN_NAME` is set) the behaviour of `card_only` stays the same: no other PAM modules should be used and so `PAM_CRED_INSUFFICIENT` is returned if no token was found.

**Comment by LudovicRousseau** [comment]
Please rebase your patches on current master first.

**Comment by wolneykien** [comment]
> Please rebase your patches on current master first.

Done.

File: src/pam_pkcs11/pam_pkcs11.c [hunk 1]
```diff
@@ -482,7 +482,8 @@ PAM_EXTERN int pam_sm_authenticate(pam_handle_t *pamh, int flags, int argc, cons
 				pam_prompt(pamh, PAM_ERROR_MSG , NULL, _("Error 2318: Empty smartcard PIN not allowed."));
 				sleep(configuration->err_display_time);
 			}
-			goto auth_failed_wrongpw;
+            pkcs11_pam_fail = PAM_AUTH_ERR;
```
Comment (LudovicRousseau) [inline src/pam_pkcs11/pam_pkcs11.c:485]: Indentation issue

File: src/pam_pkcs11/pam_pkcs11.c [hunk 2]
```diff
@@ -769,7 +772,8 @@ PAM_EXTERN int pam_sm_authenticate(pam_handle_t *pamh, int flags, int argc, cons
 			pam_prompt(pamh, PAM_ERROR_MSG , NULL, ("Error 2344: Closing PKCS#11 session failed"));
 			sleep(configuration->err_display_time);
 		}
-	goto auth_failed_wrongpw;
+    pkcs11_pam_fail = PAM_AUTH_ERR;
```
Comment (LudovicRousseau) [inline src/pam_pkcs11/pam_pkcs11.c:775]: indentation issue

File: src/pam_pkcs11/pam_pkcs11.c [hunk 3]
```diff
@@ -371,7 +371,7 @@ PAM_EXTERN int pam_sm_authenticate(pam_handle_t *pamh, int flags, int argc, cons
         /* If the login isn't restricted to card-only, then proceed
            to the next auth. module quietly. */
         release_pkcs11_module(ph);
-        goto exit_ingore;
+        goto exit_ignore;
```
Comment (LudovicRousseau) [inline src/pam_pkcs11/pam_pkcs11.c:374]: please merge this commit with the one that introduced the problem.

File: src/pam_pkcs11/pam_pkcs11.c [hunk 4]
```diff
@@ -410,58 +390,35 @@ PAM_EXTERN int pam_sm_authenticate(pam_handle_t *pamh, int flags, int argc, cons
       }
 
       if (configuration->slot_description != NULL) {
-	rv = wait_for_token_by_slotlabel(ph, configuration->slot_description,
-          login_token_name, &slot_num);
+          rv = wait_for_token_by_slotlabel(ph, configuration->slot_description,
```
Comment (LudovicRousseau) [inline src/pam_pkcs11/pam_pkcs11.c]: this patch is only indentation change.
Really needed?

File: src/pam_pkcs11/pam_pkcs11.c [hunk 5]
```diff
@@ -410,58 +390,35 @@ PAM_EXTERN int pam_sm_authenticate(pam_handle_t *pamh, int flags, int argc, cons
...
+                                           login_token_name, &slot_num);
       } else if (configuration->slot_num != -1) {
-        rv = wait_for_token(ph, configuration->slot_num,
-                          login_token_name, &slot_num);
+          rv = wait_for_token(ph, configuration->slot_num,
```
Comment (LudovicRousseau) [inline src/pam_pkcs11/pam_pkcs11.c]: this patch is only indentation change.
Really needed?

File: etc/pam_pkcs11.conf.example.in [hunk 6]
```diff
@@ -45,7 +45,7 @@ pam_pkcs11 {
   # will bypass pam_pkcs11 if a token was not used to login (The basic idea is
   # you always unlock the screen saver with the same mechanism you used to
   # login).
-  screen_savers = gnome-screensaver, kde4-kscreensaver, kscreensaver, xscreensaver;
+  screen_savers = mate-screensaver, gnome-screensaver, kde4-kscreensaver, kscreensaver, xscreensaver;
```
Comment (LudovicRousseau) [inline etc/pam_pkcs11.conf.example.in:48]: Maybe you can merge this commit in the previous one?

Comment (wolneykien) [reply etc/pam_pkcs11.conf.example.in:48]: Thanks for pointing this out. Actually, it seems reasonable to move this one to another PR: it's just an update of the screensavers list and has nothing to do with the main thing: `card_only`, `wait_for_card` and returning `PAM_IGNORE`.

File: src/pam_pkcs11/pam_pkcs11.c (hunk 1 above)
Comment (wolneykien) [reply src/pam_pkcs11/pam_pkcs11.c]: Fixed.

File: src/pam_pkcs11/pam_pkcs11.c (hunk 2 above)
Comment (wolneykien) [reply src/pam_pkcs11/pam_pkcs11.c]: Fixed.

File: src/pam_pkcs11/pam_pkcs11.c (hunk 3 above)
Comment (wolneykien) [reply src/pam_pkcs11/pam_pkcs11.c:374]: Squashed.

File: src/pam_pkcs11/pam_pkcs11.c (hunk 4 above)
Comment (wolneykien) [reply src/pam_pkcs11/pam_pkcs11.c]: As a part of a larger changes, why not?

File: src/pam_pkcs11/pam_pkcs11.c (hunk 5 above)
Comment (wolneykien) [reply src/pam_pkcs11/pam_pkcs11.c]: As a part of a larger changes, why not?

File: etc/pam_pkcs11.conf.example.in (hunk 6 above)
Comment (wolneykien) [reply etc/pam_pkcs11.conf.example.in]: Later, I've realized that the previous commit really introduces the `screen_savers` into the example config. So, you've right: we just need to join these commits. Done.

**Pull request merged**